src/v/archival/ntp_archiver_service.cc

```diff
@@ -1546,11 +1587,90 @@ ss::future<ntp_archiver::upload_group_result> ntp_archiver::wait_uploads(
+            // sent out.
+            //
+            // By doing this incrementally we can build consistent log metadata
+            // because every such check is based on previous state that was
+            // also validated using the same procedure.
+            auto stats = segment_results[i].value();
+            if (
+              upload.upload_kind == segment_upload_kind::non_compacted
+              && upload.meta.has_value()) {
+                if (
+                  upload.meta->size_bytes != stats.size_bytes
+                  || upload.meta->base_offset != stats.base_rp_offset
+                  || upload.meta->committed_offset != stats.last_rp_offset) {
+                    vlog(
+                      _rtclog.error,
+                      "Metadata of the uploaded segment [size: {}, base: {}, "
+                      "last: {}] doesn't match the segment [size: {}, base: "
+                      "{}, last: {}]",
+                      upload.meta->size_bytes,
+                      upload.meta->base_offset,
+                      upload.meta->committed_offset,
+                      stats.size_bytes,
+                      stats.base_rp_offset,
+                      stats.last_rp_offset);
+                    break;
+                }
+            }
+        }
+        if (
+          segment_kind == segment_upload_kind::non_compacted
+          && upload.meta.has_value() && last_offset.has_value()
+          && upload.meta->base_offset > last_offset.value()) {
+            // This code block is executed only for non-compacted uploads
+            // which are adding new segments and not replacing existing
+            // ones.
+            if (
+              delta_offset_top.has_value() && upload.meta.has_value()
+              && upload.meta->delta_offset != delta_offset_top) {
+                vlog(
+                  _rtclog.error,
+                  "Delta offset of the uploaded segment {} doesn't match "
+                  "with expected value of {}",
+                  upload.meta->delta_offset,
+                  delta_offset_top);
+                break;
+            } else {
+                delta_offset_top = upload.meta->delta_offset_end;
+            }
+            if (
+              last_offset.has_value() && upload.meta.has_value()
+              && upload.meta->base_offset
+                   != model::next_offset(last_offset.value())) {
+                vlog(
+                  _rtclog.error,
+                  "Base offset of the uploaded segment {} doesn't align "
+                  "with previous segment with committed offset {}",
+                  upload.meta->base_offset,
+                  model::next_offset(last_offset.value()));
+                break;
+            } else {
+                last_offset = upload.meta->committed_offset;
+            }
+        }
 
         if (segment_kind == segment_upload_kind::non_compacted) {
```

Comment:
Should we still update any of these? Perhaps the uploaded metrics? Perhaps the gap_detected metric?